mount/mount_linux.go

```diff
@@ -117,9 +170,44 @@ func (m *Mount) mount(target string) (err error) {
 		// Remount the bind to apply read only.
 		return unix.Mount("", target, "", uintptr(oflags|unix.MS_REMOUNT), "")
 	}
+
+	// remap non-overlay mount point
+	if opt.uidmap != "" && opt.gidmap != "" && m.Type != "overlay" {
+		if err := IDMapMount(target, target, usernsFd); err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
+func prepareIDMappedOverlay(lowerDirs []string, usernsFd int) (tmpLowerDirs []string, _ func(), _ error) {
+	td, err := os.MkdirTemp(tempMountLocation, "ovl-idmapped")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for i, lowerDir := range lowerDirs {
+		tmpLowerDir := filepath.Join(td, strconv.Itoa(i))
+		if err = os.MkdirAll(tmpLowerDir, 0700); err != nil {
+			return nil, nil, fmt.Errorf("failed to create temporary dir: %w", err)
+		}
+
+		if err = IDMapMount(lowerDir, tmpLowerDir, usernsFd); err != nil {
+			return nil, nil, err
+		}
+		tmpLowerDirs = append(tmpLowerDirs, tmpLowerDir)
+	}
+
+	cleanUp := func() {
```

Comment:
I think this `cleanUp` should be defined before the for-loop. If the return-error is not nil, we can clean it up as possible as we can.

Reply:
True, this way should be better, thanks!